### PR TITLE
Delete empty keytab during client installation

### PR DESCRIPTION
**Pull request by netoarmando**
Client installation fails if '/etc/krb5.keytab' exists as a zero-length file. Deleting empty keytab before proceeding with the installation fixes the problem.

https://pagure.io/freeipa/issue/7625

Signed-off-by: Armando Neto <abiagion@redhat.com>

**Note to reviewers:**
Function implemented instead of an in-place solution, so it could be easily tested, if this solution is overkill I don't mind to remove the test case and move the check-and-delete procedure to where the function is being called.

### CODE REVIEW

**Comment by Tiboris** [comment]
Thanks for the patch @netoarmando!
IMHO testing python functionality is not what we want to do.
Probably adding the test for this particular case that creates empty keytab and then performs client installation would be the test case which will actually catch the issue.

**Comment by frozencemetery** [comment]
Why does a 0-length file get created at /etc/krb5.keytab?  That's not a valid keytab file (missing header, footer, etc.).

**Comment by rcritten** [comment]
We don't know but rather than failing cryptically we figured we could detect the condition and work around it.

**Comment by netoarmando** [comment]
@Tiboris thanks for the input! I've changed an existing test case to save resources, what do you think?

@frozencemetery Do you think we should validate other properties of the keytab (header, footer, etc.) as well?

**Comment by frozencemetery** [comment]
@netoarmando If we're going to do this at all, I'd just check klist.  If `klist -kt /etc/krb5.keytab` fails (and it will for empty files), either fail out installation or remove it.

File: ipaclient/install/client.py [hunk 1]
```diff
@@ -2126,6 +2126,13 @@ def install_check(options):
         logger.warning("Option 'force-join' has no additional effect "
                        "when used with together with option 'keytab'.")
 
+    # Remove empty keytab file
+    if (
```
Comment (Tiboris) [inline ipaclient/install/client.py]: Could you please use `\` instead of `()` ?using  brackets do feel troublesome to me.
```python
if os.path.isfile(paths.KRB5_KEYTAB) and \
        os.path.getsize(paths.KRB5_KEYTAB) == 0 :
    remove_file(paths.KRB5_KEYTAB)
```

**Comment by flo-renaud** [comment]
Hi @netoarmando 
it's a good idea to plug your test into an existing test in order to save resources, but I would rather use test_installation.py as a base (it's more natural to look for a client installation test in test_installation.py rather than in test_uninstallation.py).
Also, the test is failing but I'm not sure about the test status without your change (test_uninstallation.py is not part of PRCI-triggered tests or nightly tests, meaning we have no reference for its result).

File: ipaclient/install/client.py (hunk 1 above)
Comment (netoarmando) [reply ipaclient/install/client.py]: Tibor, I think tests are failing for unrelated reasons, like Flo mentioned later, this test is not part of nightly bundle.

About breaking lines, I'm using [black](https://github.com/ambv/black) as reference, it suggests the use of parentheses in cases like this. However, after my changes this is not necessary anymore, so we are good! :-)

**Comment by netoarmando** [comment]
Thank you, @flo-renaud! Tests already defined in `test_installation.py` leverage the fact that replica installation promotes clients, testing a client specific behavior using the current test cases seemed awkward to me, that's why I chose `test_uninstallation.py`. 

But, If I understood you correctly, in this case is okay to add a new test case to `test_installation.py`?

Unrelated question: Why `test_uninstallation.py` is not included in the nightly tests?

**Comment by rcritten** [comment]
@netoarmando it was an oversight that it wasn't included in nightly.

Rather than forking out to klist you may be able to use the gssapi module to test a keytab (not sure).

**Comment by netoarmando** [comment]
@rcritten,
> Rather than forking out to klist you may be able to use the gssapi module to test a keytab (not sure).

You are right, trying to initialize a `gssapi.Credentials` object with an invalid keytab, an exception is raised: 
```
GSSError: Major (851968): Unspecified GSS failure.  \
Minor code may provide more information, Minor (2529639093): \ 
Keytab FILE:/etc/krb5.notfound is nonexistent or empty
```

However, I'm still not sure the best place to test this.

**Comment by tiran** [comment]
master:

* cf1301fb064fc230c780c4bc5eeccb723899f7b6 Delete empty keytab during client installation
